### PR TITLE
Implement Power Cap Notifications

### PR DESCRIPTION
**Pull request by amitaibu**
#630 #631 #304

### CODE REVIEW

File: negawatt/modules/custom/negawatt_normalizer/negawatt_normalizer.install [hunk 1]
```diff
@@ -0,0 +1,62 @@
+<?php
+
+/**
+ * @file
+ * @author blazey http://drupal.org/user/353861
```
Comment (amitaibu) [inline negawatt/modules/custom/negawatt_normalizer/negawatt_normalizer.install]: remove

File: negawatt/modules/custom/negawatt_normalizer/negawatt_normalizer.module [hunk 2]
```diff
@@ -120,3 +120,35 @@ function negawatt_normalizer_process_meter($node, $from_timestamp = NULL, $to_ti
   return $handler->process($frequencies = array(), $from_timestamp, $to_timestamp);
 }
 
+/**
+ * Implements hook_entity_info() for power-cap analyzer parameters table.
+ */
+function negawatt_normalizer_entity_info() {
+  $return = array(
+    'power_cap_analyzer_info' => array(
+      'label' => t('Power-cap analyzer parameters', array(), array('context' => 'power-cap analyzer info entity')),
```
Comment (amitaibu) [inline negawatt/modules/custom/negawatt_normalizer/negawatt_normalizer.module]: `t('Power-cap analyzer parameters'),`

**Comment by amitaibu** [comment]
Check https://www.drupal.org/project/eck

File: negawatt/modules/custom/negawatt_normalizer/negawatt_normalizer.module [hunk 3]
```diff
@@ -120,3 +120,35 @@ function negawatt_normalizer_process_meter($node, $from_timestamp = NULL, $to_ti
...
 }
 
+/**
+ * Implements hook_entity_info() for power-cap analyzer parameters table.
+ */
+function negawatt_normalizer_entity_info() {
+  $return = array(
+    'power_cap_analyzer_info' => array(
+      'label' => t('Power-cap analyzer parameters'),
+      'locking mode' => 'pessimistic',
```
Comment (amitaibu) [inline negawatt/modules/custom/negawatt_normalizer/negawatt_normalizer.module]: remove

File: negawatt/modules/custom/negawatt_normalizer/negawatt_normalizer.module [hunk 4]
```diff
@@ -120,3 +120,35 @@ function negawatt_normalizer_process_meter($node, $from_timestamp = NULL, $to_ti
...
+
+  return $return;
+}
+
+/**
+ * Returns power-cap analyzer info controller.
+ *
+ * @return PowerCapAnalyzerInfoEntityController
+ */
+function negawatt_power_cap_analyzer_info_get_controller() {
```
Comment (amitaibu) [inline negawatt/modules/custom/negawatt_normalizer/negawatt_normalizer.module]: is it needed?

File: negawatt/modules/custom/negawatt_normalizer/plugins/negawatt_analyzer/NegaWattAnalyzerPowerCap.php [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+   */
+  public function process(array $data, $create_message = TRUE) {
+    $meter_nid = 0;
+    $return = array();
+    $entities_of_frequency = $data[\NegaWattNormalizerTimeManagerInterface::MINUTE];
+    foreach ($entities_of_frequency as $timestamp => $entities_of_timeframe) {
+
+      foreach ($entities_of_timeframe as $rate_type => $entity) {
+
+        $weekday = date('w', $entity->timestamp) + 1; // Sun = 1 .. Sat = 7
```
Comment (amitaibu) [inline negawatt/modules/custom/negawatt_normalizer/plugins/negawatt_analyzer/NegaWattAnalyzerPowerCap.php]: comment above

File: negawatt/modules/custom/negawatt_normalizer/plugins/negawatt_analyzer/NegaWattAnalyzerPowerCap.php [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+          ->propertyCondition('weekday', $weekday)
+          ->propertyCondition('hour_from', $hour, '<=')
+          ->propertyCondition('hour_to', $hour, '>');
+
+        $result = $query->execute();
+
+        // If prev year's record was found, check that it does not differ by more than 10%
+        if (!empty($result)) {
+          // @fixme: in case of multiple matches, take minimal cap value.
+          $eids = array_keys($result['power_cap_analyzer_info']);
```
Comment (amitaibu) [inline negawatt/modules/custom/negawatt_normalizer/plugins/negawatt_analyzer/NegaWattAnalyzerPowerCap.php]: => `$entity_ids`

File: negawatt/modules/custom/negawatt_normalizer/plugins/negawatt_analyzer/NegaWattAnalyzerPowerCap.php [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+          $entities = entity_load('power_cap_analyzer_info', $eids);
+          $cap = $entities[$eids[0]]->cap;
+
+          if ($entity->avg_power > $cap) {
+            // Average power consumed is higher that the cap, put an alert
+            // Prepare message parameters.
+            $return[] = array(
+              'nid' => $meter_nid,
+              'message_type' => 'anomalous_consumption',
```
Comment (amitaibu) [inline negawatt/modules/custom/negawatt_normalizer/plugins/negawatt_analyzer/NegaWattAnalyzerPowerCap.php]: every analyzer should have its own message type

File: negawatt/modules/custom/negawatt_normalizer/tests/NegawattNormalizerMeterProcessTestCase.test [hunk 8]
```diff
@@ -367,9 +378,20 @@ class NegawattNormalizerMeterProcessTestCase extends NegawattWebTestCase {
     // Problem with float representation prevent the exact value of 0.96
     $this->assertTrue(abs($result_entity->min_power_factor - 0.96) < 0.0001, 'Min power-factor calculation was correct.');
 
-    // Verify the calculation of the first normalized entity
+    // Verify the calculation of the first normalized entity - the avg_power should be null.
     $result_entity = $result_entities[0];
-    $this->assertTrue($result_entity->avg_power < 0.0001, 'Average power calculation for first entity was correct.');
+    $this->assertNull($result_entity->avg_power, 'Average power calculation for first entity was correct.');
+
+    // Make sure a notification message was generated.
```
Comment (amitaibu) [inline negawatt/modules/custom/negawatt_normalizer/tests/NegawattNormalizerMeterProcessTestCase.test]: move to own test

File: negawatt/modules/custom/negawatt_normalizer/plugins/negawatt_analyzer/NegaWattAnalyzerPowerCap.php [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+      foreach ($entities_of_timeframe as $rate_type => $entity) {
+
+        $weekday = date('w', $entity->timestamp) + 1; // Sun = 1 .. Sat = 7
+        $hour = date('H', $entity->timestamp);
+        if (empty($meter_nid)) {
+          $meter_nid = $entity->meter_nid;
+        }
+        // Look for the relevant power-cap entity
+        $query = new EntityFieldQuery();
+        $query->entityCondition('entity_type', 'power_cap_analyzer_info')
```
Comment (amitaibu) [inline negawatt/modules/custom/negawatt_normalizer/plugins/negawatt_analyzer/NegaWattAnalyzerPowerCap.php]: `$result = ...`

File: negawatt/modules/custom/negawatt_normalizer/plugins/negawatt_analyzer/NegaWattAnalyzerPowerCap.php [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+          ->propertyCondition('meter_nid', $meter_nid)
+          ->propertyCondition('weekday', $weekday)
+          ->propertyCondition('hour_from', $hour, '<=')
+          ->propertyCondition('hour_to', $hour, '>');
+
+        $result = $query->execute();
+
+        // If prev year's record was found, check that it does not differ by more than 10%
+        if (!empty($result)) {
```
Comment (amitaibu) [inline negawatt/modules/custom/negawatt_normalizer/plugins/negawatt_analyzer/NegaWattAnalyzerPowerCap.php]: `if (!empty($result['power_cap_analyzer_info')`

**Comment by shahardo** [comment]
Merged

**Pull request closed without merging**